### DISCUSSION TITLE
createRenderRoot with custom target does not remove component on disconnectedCallback

### DISCUSSION THREAD

**Discussion by steve2507** [lit/lit#4825] [Q&A] [answered]
I have been trying to implement a modal component which simply targets a root node somewhere using `createRenderRoot`:
```ts
  createRenderRoot() {
    return document.querySelector('#overlays');
  }
```

This works when _adding_ the component. However, whenever `disconnectedCallback` is called, the component is not removed.
Is this by design?

A repro can be found [here](https://lit.dev/playground/#project=W3sibmFtZSI6InNpbXBsZS1ncmVldGluZy50cyIsImNvbnRlbnQiOiJpbXBvcnQge2h0bWwsIGNzcywgTGl0RWxlbWVudH0gZnJvbSAnbGl0JztcbmltcG9ydCB7Y3VzdG9tRWxlbWVudCwgcHJvcGVydHl9IGZyb20gJ2xpdC9kZWNvcmF0b3JzLmpzJztcblxuQGN1c3RvbUVsZW1lbnQoJ3NpbXBsZS1ncmVldGluZycpXG5leHBvcnQgY2xhc3MgU2ltcGxlR3JlZXRpbmcgZXh0ZW5kcyBMaXRFbGVtZW50IHtcbiAgc3RhdGljIHN0eWxlcyA9IGNzc2BwIHsgY29sb3I6IGJsdWUgfWA7XG5cbiAgQHByb3BlcnR5KClcbiAgbmFtZSA9ICdTb21lYm9keSc7XG4gIFxuICBjcmVhdGVSZW5kZXJSb290KCkge1xuICAgIHJldHVybiBkb2N1bWVudC5xdWVyeVNlbGVjdG9yKCcjb3ZlcmxheXMnKTtcbiAgfVxuXG4gIHJlbmRlcigpIHtcbiAgICByZXR1cm4gaHRtbGA8cD5IZWxsbywgJHt0aGlzLm5hbWV9ITwvcD5gO1xuICB9XG59XG4ifSx7Im5hbWUiOiJpbmRleC5odG1sIiwiY29udGVudCI6IjwhRE9DVFlQRSBodG1sPlxuPGhlYWQ-XG4gIDxzY3JpcHQgdHlwZT1cIm1vZHVsZVwiIHNyYz1cIi4vc2ltcGxlLXRvZ2dsZS5qc1wiPjwvc2NyaXB0PlxuPC9oZWFkPlxuPGJvZHk-XG4gIDxwPlxuICAgIENsaWNrIHRoZSBidXR0b24gdG8gdG9nZ2xlIHRoZSBncmVldGluZy4gT2JzZXJ2ZSB0aGF0IHRoZSBncmVldGluZyBzdGF5cyB3aGVuIHRvZ2dsaW5nIG9mZi5cbiAgPC9wPlxuICA8ZGl2IGlkPVwib3ZlcmxheXNcIj48L2Rpdj5cbiAgPHNpbXBsZS10b2dnbGU-PC9zaW1wbGUtdG9nZ2xlPlxuPC9ib2R5PlxuIn0seyJuYW1lIjoicGFja2FnZS5qc29uIiwiY29udGVudCI6IntcbiAgXCJkZXBlbmRlbmNpZXNcIjoge1xuICAgIFwibGl0XCI6IFwiXjMuMC4wXCIsXG4gICAgXCJAbGl0L3JlYWN0aXZlLWVsZW1lbnRcIjogXCJeMi4wLjBcIixcbiAgICBcImxpdC1lbGVtZW50XCI6IFwiXjQuMC4wXCIsXG4gICAgXCJsaXQtaHRtbFwiOiBcIl4zLjAuMFwiXG4gIH1cbn0iLCJoaWRkZW4iOnRydWV9LHsibmFtZSI6InNpbXBsZS10b2dnbGUudHMiLCJjb250ZW50IjoiaW1wb3J0IHtodG1sLCBjc3MsIExpdEVsZW1lbnR9IGZyb20gJ2xpdCc7XG5pbXBvcnQge2N1c3RvbUVsZW1lbnQsIHByb3BlcnR5LCBzdGF0ZX0gZnJvbSAnbGl0L2RlY29yYXRvcnMuanMnO1xuXG5pbXBvcnQgJy4vc2ltcGxlLWdyZWV0aW5nLmpzJztcblxuQGN1c3RvbUVsZW1lbnQoJ3NpbXBsZS10b2dnbGUnKVxuZXhwb3J0IGNsYXNzIFNpbXBsZVRvZ2dsZSBleHRlbmRzIExpdEVsZW1lbnQge1xuICBzdGF0aWMgc3R5bGVzID0gY3NzYHAgeyBjb2xvcjogYmx1ZSB9YDtcblxuICBAcHJvcGVydHkoKVxuICBuYW1lID0gJ1NvbWVib2R5JztcbiAgXG4gIEBzdGF0ZSgpXG4gIHZpc2libGUgPSBmYWxzZTtcblxuICByZW5kZXIoKSB7XG4gICAgcmV0dXJuIGh0bWxgXG4gICAgICA8YnV0dG9uXG4gICAgICAgIEBjbGljaz0keygpID0-IHt0aGlzLnZpc2libGUgPSAhdGhpcy52aXNpYmxlfX1cbiAgICAgID5cbiAgICAgICAgVG9nZ2xlICR7dGhpcy52aXNpYmxlfVxuICAgICAgPC9idXR0b24-XG4gICAgICAke3RoaXMudmlzaWJsZVxuICAgICAgICA_IGh0bWxgXG4gICAgICAgICAgPHNpbXBsZS1ncmVldGluZyBuYW1lPVwiV29ybGRcIiAvPlxuICAgICAgICBgXG4gICAgICAgIDogJydcbiAgICAgIH1cbiAgICBgO1xuICB9XG59XG4ifV0).

Thank you for your help.

**Comment by augustjk** [comment, marked as answer]
I believe this is working as intended. Lit won't touch anything outside of itself. Specifying the render root to be outside the element can get tricky as it's easier for some other agent to be modifying that element too.

For implementing a modal component, rather than conditionally rendering a component that targets a render root outside of itself, it would be better place a modal element where you want and either get a handle to it and imperatively set properties on it, or fire an event that the modal component can listen for to update itself.

If you must specify a different render root outside the element like that, you should be responsible for cleaning it up too, which can be done by overriding `disconnectedCallback()` like [this](https://lit.dev/playground/#project=W3sibmFtZSI6InNpbXBsZS1ncmVldGluZy50cyIsImNvbnRlbnQiOiJpbXBvcnQge2h0bWwsIGNzcywgTGl0RWxlbWVudCwgcmVuZGVyLCBub3RoaW5nfSBmcm9tICdsaXQnO1xuaW1wb3J0IHtjdXN0b21FbGVtZW50LCBwcm9wZXJ0eX0gZnJvbSAnbGl0L2RlY29yYXRvcnMuanMnO1xuXG5AY3VzdG9tRWxlbWVudCgnc2ltcGxlLWdyZWV0aW5nJylcbmV4cG9ydCBjbGFzcyBTaW1wbGVHcmVldGluZyBleHRlbmRzIExpdEVsZW1lbnQge1xuICBzdGF0aWMgc3R5bGVzID0gY3NzYHAgeyBjb2xvcjogYmx1ZSB9YDtcblxuICBAcHJvcGVydHkoKVxuICBuYW1lID0gJ1NvbWVib2R5JztcbiAgXG4gIGNyZWF0ZVJlbmRlclJvb3QoKSB7XG4gICAgcmV0dXJuIGRvY3VtZW50LnF1ZXJ5U2VsZWN0b3IoJyNvdmVybGF5cycpIGFzIEhUTUxFbGVtZW50O1xuICB9XG4gIFxuICBkaXNjb25uZWN0ZWRDYWxsYmFjaygpIHtcbiAgICByZW5kZXIobm90aGluZywgZG9jdW1lbnQucXVlcnlTZWxlY3RvcignI292ZXJsYXlzJykgYXMgSFRNTEVsZW1lbnQpO1xuICAgIHN1cGVyLmRpc2Nvbm5lY3RlZENhbGxiYWNrKCk7XG4gIH1cblxuICByZW5kZXIoKSB7XG4gICAgcmV0dXJuIGh0bWxgPHA-SGVsbG8sICR7dGhpcy5uYW1lfSE8L3A-YDtcbiAgfVxufVxuIn0seyJuYW1lIjoiaW5kZXguaHRtbCIsImNvbnRlbnQiOiI8IURPQ1RZUEUgaHRtbD5cbjxoZWFkPlxuICA8c2NyaXB0IHR5cGU9XCJtb2R1bGVcIiBzcmM9XCIuL3NpbXBsZS10b2dnbGUuanNcIj48L3NjcmlwdD5cbjwvaGVhZD5cbjxib2R5PlxuICA8cD5cbiAgICBDbGljayB0aGUgYnV0dG9uIHRvIHRvZ2dsZSB0aGUgZ3JlZXRpbmcuIE9ic2VydmUgdGhhdCB0aGUgZ3JlZXRpbmcgc3RheXMgd2hlbiB0b2dnbGluZyBvZmYuXG4gIDwvcD5cbiAgPGRpdiBpZD1cIm92ZXJsYXlzXCI-PC9kaXY-XG4gIDxzaW1wbGUtdG9nZ2xlPjwvc2ltcGxlLXRvZ2dsZT5cbjwvYm9keT5cbiJ9LHsibmFtZSI6InBhY2thZ2UuanNvbiIsImNvbnRlbnQiOiJ7XG4gIFwiZGVwZW5kZW5jaWVzXCI6IHtcbiAgICBcImxpdFwiOiBcIl4zLjAuMFwiLFxuICAgIFwiQGxpdC9yZWFjdGl2ZS1lbGVtZW50XCI6IFwiXjIuMC4wXCIsXG4gICAgXCJsaXQtZWxlbWVudFwiOiBcIl40LjAuMFwiLFxuICAgIFwibGl0LWh0bWxcIjogXCJeMy4wLjBcIlxuICB9XG59IiwiaGlkZGVuIjp0cnVlfSx7Im5hbWUiOiJzaW1wbGUtdG9nZ2xlLnRzIiwiY29udGVudCI6ImltcG9ydCB7aHRtbCwgY3NzLCBMaXRFbGVtZW50fSBmcm9tICdsaXQnO1xuaW1wb3J0IHtjdXN0b21FbGVtZW50LCBwcm9wZXJ0eSwgc3RhdGV9IGZyb20gJ2xpdC9kZWNvcmF0b3JzLmpzJztcblxuaW1wb3J0ICcuL3NpbXBsZS1ncmVldGluZy5qcyc7XG5cbkBjdXN0b21FbGVtZW50KCdzaW1wbGUtdG9nZ2xlJylcbmV4cG9ydCBjbGFzcyBTaW1wbGVUb2dnbGUgZXh0ZW5kcyBMaXRFbGVtZW50IHtcbiAgc3RhdGljIHN0eWxlcyA9IGNzc2BwIHsgY29sb3I6IGJsdWUgfWA7XG5cbiAgQHByb3BlcnR5KClcbiAgbmFtZSA9ICdTb21lYm9keSc7XG4gIFxuICBAc3RhdGUoKVxuICB2aXNpYmxlID0gZmFsc2U7XG5cbiAgcmVuZGVyKCkge1xuICAgIHJldHVybiBodG1sYFxuICAgICAgPGJ1dHRvblxuICAgICAgICBAY2xpY2s9JHsoKSA9PiB7dGhpcy52aXNpYmxlID0gIXRoaXMudmlzaWJsZX19XG4gICAgICA-XG4gICAgICAgIFRvZ2dsZSAke3RoaXMudmlzaWJsZX1cbiAgICAgIDwvYnV0dG9uPlxuICAgICAgJHt0aGlzLnZpc2libGVcbiAgICAgICAgPyBodG1sYFxuICAgICAgICAgIDxzaW1wbGUtZ3JlZXRpbmcgbmFtZT1cIldvcmxkXCIgLz5cbiAgICAgICAgYFxuICAgICAgICA6ICcnXG4gICAgICB9XG4gICAgYDtcbiAgfVxufVxuIn1d).
